### ISSUE TITLE
Custom Dataset

### ISSUE THREAD

**Issue by pytholic** [dornik/reagent#1]
Hi. Thank you for the repo!
I wanted to ask if you can guide a bit on making dataloader for custom dataset. My data contains ".obj" files.

**Comment by dornik** [comment]
Hi!

You probably first want to import your object models and generate point clouds. Please see the [data preparation](https://github.com/dornik/reagent/blob/main/dataset/dataset.py#L225) in `DatasetLinemod` on how to use [`trimesh`](https://trimsh.org/index.html) for this. Don't forget to zero-center the point clouds and normalize them to [-1,1].

Now, if your goal is to align two point cloud instances of a certain model (along the lines of the ModelNet40 experiments in the paper), you can use the [`SplitSourceRef`](https://github.com/dornik/reagent/blob/f6497af6c62eadf8e34245cf6c2d9f87508c990b/dataset/augmentation.py#L40) transformation to generate a pair of source/target point clouds. The [`getitem`](https://github.com/dornik/reagent/blob/main/dataset/dataset.py#L35) method in `DatasetModelNet40` shows how a minimal `sample` dictionary should look like. You add one of the previously generated model point clouds as `points`, a running sample index as `idx` (set as random seed during evaluation for reproducibility) and, optionally, the object category as `label` for evaluation.

Finally, you have to define the transformations that should be applied to each sample. You might want to add random noise to the point clouds during training or scale them randomly within a certain range. Please take a look at the [`get_transforms`](https://github.com/dornik/reagent/blob/main/dataset/dataset.py#L42) methods and the respective [transformations](https://github.com/dornik/reagent/blob/f6497af6c62eadf8e34245cf6c2d9f87508c990b/dataset/augmentation.py) themselves. This will require some tuning depending on your specific dataset to get a good performance.

Hope this clarifies things!

**Comment by pytholic** [comment]
@dornik Hi and thank you for your detailed response. I was able to work with my data using open3d and custom dataloader (by following your snippet). Thank you for the repo again. Cheers!

**Comment by pytholic** [comment]
@dornik Although I am curios about another thing. I am not sure if you are much familiar with open3d, I am also kinda new to it. But I was thinking about visualizing results with open3d somehow instead of 2D visualization. You think it is feasible?

**Comment by dornik** [comment]
Should be straightforward to create a new visualizer class [here](https://github.com/dornik/reagent/blob/main/utility/visualization.py) that uses Open3D's visualization tools. 

Following the naming convention of my existing visualizers, initialize the O3D window and add your `PointCloud` geometries in `reset` like [here](http://www.open3d.org/docs/0.9.0/tutorial/Advanced/non_blocking_visualization.html#initialize-visualizer-class). Then you simply update the geometries by setting, e.g., the `source.points` variable in `update`. That way you shouldn't have to change too much in the evaluation script either.

**Comment by pytholic** [comment]
@dornik thanks a lot for all the help. I made the open3d visualization work with modifications in your script. I will close the issue now :)

**Comment by dornik** [comment]
Sure, no worries. Glad it worked :smile:
